### ISSUE TITLE
None of the “sha512” hashes in the integrity attribute match the content of the subresource

### ISSUE THREAD

**Issue by nebulade** [PrivateBin/PrivateBin#1170]
After update to 1.6.0 opening the app, the browser shows the following error and the app cannot be used. chrome and firefox show slightly different error:

chrome:
```
Failed to find a valid digest in the 'integrity' attribute for resource 'https://<domain>/js/privatebin.js?1.6.0' with computed SHA-512 integrity 'bvLSauH52p3a1alLFQ7YYbl95KUHaf4oVPsUeIqVEBry0c30By+chwu3o5cXUPWB/+OAz0TY00P+k+lquMsAcQ=='. The resource has been blocked.
```

firefox:
```
None of the “sha512” hashes in the integrity attribute match the content of the subresource. The computed hash is “bvLSauH52p3a1alLFQ7YYbl95KUHaf4oVPsUeIqVEBry0c30By+chwu3o5cXUPWB/+OAz0TY00P+k+lquMsAcQ==”.
```

Any hint to debug this further is highly appreciated or some pointers on what the error might actually mean in the private bin context so we can start investigating at least in the right direction.

I have seen one issue which was related to Cloudflare tampering with the content, but in this case there is no cloudflare involved. https://github.com/PrivateBin/PrivateBin/issues/453 and https://github.com/PrivateBin/PrivateBin/issues/98

### What happens

The UI only  then shows:
```
In case this message never disappears please have a look at this FAQ for information to troubleshoot.
```

## Basic information

**Server OS:** Cloudron

**PrivateBin version:** 1.6.0

I can reproduce this issue on <https://privatebin.net>: No

**Comment by rugk** [comment]
Does [this FAQ question](https://github.com/PrivateBin/PrivateBin/wiki/FAQ#why-does-the-loading-message-not-go-away) or the other ones following answer your question – the error message already pointed you there?
If not, please comment on why. If so, feel free to improve the FAQ and make it more clear, in case that was an issue.

Usual culprits are also listed there, such as wrong line endings when copying files (especially when switching/transferring from/to Windows/Linux etc.).

**Comment by nebulade** [comment]
I see, so this seems to be an issue of the files maybe being corrupt or tampered with? Is there a way to validate the hash via commandline to hunt down the issue? In this case this is a deployment via Cloudron, which has all files actually read-only from a Dockerimage which works otherwise fine, so is it possible that those changed files may be located in custom_templates ?

**Comment by rugk** [comment]
If the files are not modified everything should work.

See https://github.com/PrivateBin/PrivateBin/wiki/Development#subresource-integrity-for-javascript-resources for how to generate the hashes manually. It won't really help though as you just see different hash values.

Better idea is to compare files with any diff tool of for choice such as Linux diff. See e.g. https://en.m.wikipedia.org/wiki/Comparison_of_file_comparison_tools

Custom templates do not matter unless you use one and have modified the hashes there.

**Comment by elrido** [comment]
Well, if there is a custom template being used, than that template would need to get updated with the new hashes.

I tried figuring out where that tpl/custom.php comes from in the Cloudron repo, but I'm not familiar with how that hoster works and failed to spot it. I would recommend to raise an issue in the Cloudron repo since this may be something they need to handle on their end: https://git.cloudron.io/cloudron/privatebin-app/-/issues

**Comment by nebulade** [comment]
Ah yes, I am one of the Cloudron authors and I am handling a user issue, which is why I came here in the first place to better understand what that file hashing does. I can see the user having put a custom template so that seems to be the root cause, only that it works fine without modification in v1.5.2 but not v1.6.0.

So the template custom.php in this case sources js files which have an `integrity` attribte, is my assumption right, that this is what is checked against and the update may have updated files like `js/privatebin.js` which now need updateing of that attribute?

**Comment by elrido** [comment]
These hashes are not a property of the js files, but a hashsum over those files.

The user will have to update their custom.php files hashes to match those of the project on every update or they can:
1. Generate the hashes using one of the methods described under this link and update their template accordingly: https://github.com/PrivateBin/PrivateBin/wiki/Development#subresource-integrity-for-javascript-resources
2. Or they can just peek into any of the templates that shipped with 1.6.0 - they will be the same for all templates:
https://github.com/PrivateBin/PrivateBin/blob/bbc5b627c208583e4631326dadb43b451d6c8f10/tpl/bootstrap.php#L45-L76
3. Or they could even choose to remove these hashes, if they find that too cumbersome - that would of course weaken their instances security so I would not recommend doing that.

**Comment by nebulade** [comment]
Thanks for the all detailed description and it shows that this is clearly not an upstream issue but expected behavior. Will close this issue then and inform the user accordingly.
